### ISSUE TITLE
Firebase Auth fails in SwiftUI App init

### ISSUE THREAD

**Issue by otymartin** [firebase/firebase-ios-sdk#10200]
### Description

`PhoneAuthProvider.provider().verifyPhoneNumber....` will fail to execute. No success callback, no error either.

### Reproducing the issue

On Xcode 14, While using the SwiftUI lifecyle, if you configure firebase in the `@Main init()` then Firebase auth wont work.
```swift 
@Main 
struct AppName: App { 
   @UIApplicationDelegateAdaptor(AppDelegate.self) var appDelegate
    init() {
       FirebaseApp.configure()
       // This init method runs before `didFinishLaunchingWithOptions` in Xcode 14
    }
  }
```

The reason I even configured Firebase in the init method is because configuring it in the `AppDelegate's` `didFinishLaunchingWithOptions` method would crash my app since the `@Main init()` method runs first and I had firebase methods running in there before `didFinishLaunchingWithOptions` even executed meaning I was running firebase database functions before `FirebaseApp.configure()` ran.

So Running `FirebaseApp.configure()` in the `@Main init()` method would result in `PhoneAuthProvider.provider().verifyPhoneNumber` not working at all. no callback, no error nothing. But Running it in `didFinishLaunchingWithOptions` would work providing I ran my firebase database methods after a publisher told me it was safe to do so.

My temporary fix.

```swift
@MainActor
class AppDelegateObserver: ObservableObject {
    @Published var didFinishLaunching = false
    static var shared: AppDelegateObserver { AppDelegateObserver() }
}

class AppDelegate: UIResponder, UIApplicationDelegate, UNUserNotificationCenterDelegate, OSSubscriptionObserver {
    
    func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {
        FirebaseApp.configure()
        AppDelegateObserver.shared.didFinishLaunching = true
        return true
    }
    
    @Main 
    init() {
    ...
    AppDelegateObserver
            .shared
            .$didFinishLaunching
            .receive(on: RunLoop.main)
            .sink { [weak self] didFinishLaunching in
                /// Execute firebase database methods
            }
            .store(in: &disposables)
   ```
   
   Ideally, I'd rather Configure Firebase in the `@Main init()` method than resort to this bandaid fix.

Hope that helps anyone who runs into this same issue. I tested with previous versions of Firebase eg. 9.4.1, 9.3.0 and the issue remained.

### Firebase SDK Version

9.5.0

### Xcode Version

14

### Installation Method

Swift Package Manager

### Firebase Product(s)

Analytics, Authentication, Crashlytics, DynamicLinks, Firestore, Functions

### Targeted Platforms

iOS

### If using Swift Package Manager, the project's Package.resolved

<details>
<summary>Expand <code>Package.resolved</code> snippet</summary>
<br>

```json

Replace this line with the contents of your Package.resolved.

```

</details>

### If using CocoaPods, the project's Podfile.lock

<details>
<summary>Expand <code>Podfile.lock</code> snippet</summary>
<br>

```yml

Replace this line with the contents of your Podfile.lock!

```

</details>

**Comment by rizafran** [comment]
Hi @otymartin, it seems like adding the `FirebaseApp.configure()` to `@Main init()` is not sufficient as Firebase phone auth uses method swizzling to hook into the application life cycle. You may check this [article](https://peterfriese.dev/posts/swiftui-new-app-lifecycle-firebase/) for more info. With this, you need to use `@UIApplicationDelegateAdaptor` property wrapper to connect your app to an instance of `AppDelegate`, and I think you still need to call `FirebaseApp.configure()` in `didFinishLaunchingWithOptions`. Could you try and see if it works?

**Comment by paulb777** [comment]
@otymartin Would you provide some context on why Authentication needs to be initialized before `didFinishLaunching`?

**Comment by otymartin** [comment]
@paulb777 
In the new `SwiftUI` lifecycle `iOS 16+ Target` `Xcode 14`

```swift
@Main
struct MyApp: App {
 @StateObject var viewModel = MainViewModel()
  init() {
    /// ⚠️This init runs before `didFinishLaunching`
    /// ⚠️I perform auth here when `MainViewModel()` initializes
    /// ⚠️The app will crash trying to use Firebase methods before `didFinishLaunching`
  }
  
   var body: some Scene {
    WindowGroup {
      switch viewModel.state {
        case loggedIn:
         ContentView()
        case loggedOut:
         LoginView()
      }
    }
   }
}
```

**Comment by forrestsyrett** [comment]
@otymartin Were you able to find any sort of fix for this issue? It seems attempting to initialize any environmentObject that references Firestore will cause a crash because `@main init()` is called before `didFinishLaunching` in iOS 16. 

In versions prior to iOS 16 calling `FirebaseApp.configure()` from the `AppDelegate` and using `@UIApplicationDelegateAdaptor(AppDelegate.self) var appDelegate` was sufficient to meet the requirements for FirebaseMessaging and registering for remote notifications, while still initializing any environmentObjects needed at the launch of the app.

**Comment by onisiforos7** [comment]
Any update on this? @otymartin

**Comment by otymartin** [comment]
@forrestsyrett @onisiforos7 As I documented above, my solution was to have a trigger in `didFinishLaunchingWithOptions` that will tell me when the app has launched, then I can run my firebase methods in my main ViewModel's init method.

```swift 
@MainActor
class AppDelegateObserver: ObservableObject {
    @Published var didFinishLaunching = false
    static var shared: AppDelegateObserver { AppDelegateObserver() }
}

class AppDelegate: UIResponder, UIApplicationDelegate, UNUserNotificationCenterDelegate, OSSubscriptionObserver {
    
    func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {
        FirebaseApp.configure()
        AppDelegateObserver.shared.didFinishLaunching = true
        return true
    }
}

  @Main 
  struct My_App: App {
    @StateObject var viewModel = ViewModel()
  }
  
  class ViewModel: ObservableObject {
      var disposables = Set<AnyCancellable>()
       
      init() {
            AppDelegateObserver
                    .shared
                    .$didFinishLaunching
                    .receive(on: RunLoop.main)
                    .sink { [weak self] didFinishLaunching in
                        /// Execute firebase database methods
                       // db.collection("Documents").document("id").getDocument() 
                    }
                    .store(in: &disposables)
                    }
        }
}

**Comment by onisiforos7** [comment]
> ```swift
> AppDelegateObserver.shared.didFinishLaunching = true
> ```

Thank you for your answer. Do you by any chance use App Check? it looks like its not working for e on iOS 16

**Comment by onisiforos7** [comment]
@otymartin 
Thank you for your answer. Do you by any chance use App Check? it looks like its not working for me on iOS 16

**Comment by Anteloper** [comment]
Thanks for this @otymartin - is this still the best solution we know of? Architecting a new app with iOS 16 Xcode 14 and experiencing the same.

**Comment by atultw** [comment]
@otymartin Just spent a whole day fixing the same issue. Dead silence, no errors or warnings. It did indeed have to do with the placement of `FirebaseApp.configure()`. 

 The architecture of the iOS sdk is problematic. Global instances everywhere make the whole setup precarious.

**Comment by charlotteliang** [comment]
@otymartin can you send me the original code that's not working for you? Specifically I want to see what firebase database methods you are calling inside init method that cause the crash. 
Feel free to send me your sample project at chliang at google dot com.

**Comment by charlotteliang** [comment]
Generally you should call all firebase methods either in `init()` or `didFinishLaunchingWithOptions `. It's not recommended to call FirebaseApp.configure() in one and the rest in the other, because we can't control the order they are triggered by Apple system, but we can control the order if you call them all in a serial order inside one of those methods above and ensure FirebaseApp.configure() is the first one you call before other Firebase methods.

Can you try that and let me know if it's not working for you? @atultw, @otymartin

**Comment by charlotteliang** [comment]
@otymartin I'm not able to reproduce the issue. I called FirebaseApp.configure in the init() function and Phone Auth works as expected.  

Keep in mind that swizzling doesn't work in SwiftUI, so you will have to manually  implement silent notification methods  as shown in the instruction "[Appendix: Using phone sign-in without swizzling](https://firebase.google.com/docs/auth/ios/phone-auth#appendix:-using-phone-sign-in-without-swizzling)".

**Comment by forrestsyrett** [comment]
@charlotteliang The issue for me was that I had firebase messaging setup using `@UIApplicationDelegateAdaptor` so that messaging was registered, and then I had my own ObservableObject classes that referenced Firebase being initialized within the new lifecycle init(). That way my views that needed to reference these environment objects had them at launch. 
Previous to iOS 16/Xcode 14, this worked fine. The update changed this order of initialization it seems which causes a crash within the observable object classes because Firebase isn’t initialized yet.

**Comment by charlotteliang** [comment]
@forrestsyrett Your scenario is a separate issue than the original post. Would you mind filing a separate issue so we can handle it more clearly for both routes.

@atultw Would you mind send your sample project to my email chliang at google dot com, I'm guessing it's because swizzling works in SceneDelegate in SwiftUI, but like to confirm.

**Comment by forrestsyrett** [comment]
@charlotteliang the base issue is the same as @atultw explained. Splitting up the initialization between the app delegate and @Main init causes a crash because of the execution order in Xcode 14. 

> The reason I even configured Firebase in the init method is because configuring it in the AppDelegate's didFinishLaunchingWithOptions method would crash my app since the @Main init() method runs first and I had firebase methods running in there before didFinishLaunchingWithOptions even executed meaning I was running firebase database functions before FirebaseApp.configure() ran.

**Comment by charlotteliang** [comment]
@atultw Did you manually implement silent push notifications handling as I mentioned [here](https://github.com/firebase/firebase-ios-sdk/issues/10200#issuecomment-1420287656). Otherwise phone auth will not work because swizzling is not working properly in SwiftUI.

@forrestsyrett can you send me your code sample? I'm not able to reproduce the crash.

**Comment by atultw** [comment]
@charlotteliang I did, as shown in the code. Mine is working fine I just wanted to share in case anyone else experiences this.

**Comment by thedore17** [comment]
@atultw Can you please share your complete sample code?

I am running into the same problem. I cannot get Firebase to configure either in init or app delegate. As soon as I try the phone auth method, I got a crash telling me to run app configure.

This is a pretty big problem that one of the most popular verification methods for Firebase is completely broken in SwiftUI consider we are too years into the SwiftUI migration now.

**Comment by thedore17** [comment]
Here's my code. Xcode 14.2, iOS 16.2. Crashes with the message:

> The default FirebaseApp instance must be configured before the default Authinstance can be initialized.

I get that whether I call configure() in the AppDelegate or in the Init method as soon as I call PhoneAuthProvider.provider().verifyPhoneNumber in the view.

Here's my code:

`import SwiftUI
import FirebaseCore
import Firebase


class AppDelegate: NSObject, UIApplicationDelegate {
    func application(_ application: UIApplication,
                   didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey : Any]? = nil) -> Bool {
        FirebaseApp.configure()
        return true
    }
    
    func application(_ application: UIApplication, didRegisterForRemoteNotificationsWithDeviceToken deviceToken: Data) {
      // Pass device token to auth
        Auth.auth().setAPNSToken(deviceToken, type: .prod)

      // Further handling of the device token if needed by the app
      // ...
    }
    
    func application(_ application: UIApplication,
        didReceiveRemoteNotification notification: [AnyHashable : Any],
        fetchCompletionHandler completionHandler: @escaping (UIBackgroundFetchResult) -> Void) {
      if Auth.auth().canHandleNotification(notification) {
        completionHandler(.noData)
        return
      }
      // This notification is not auth related; it should be handled separately.
    }
    
    func application(_ application: UIApplication, open url: URL,
                     options: [UIApplication.OpenURLOptionsKey : Any]) -> Bool {
      if Auth.auth().canHandle(url) {
        return true
      } else {
          // if unrelated push, do something else
          // else return
          return false
      }
      // URL not auth related; it should be handled separately.
    }
    
    func scene(_ scene: UIScene, openURLContexts URLContexts: Set<UIOpenURLContext>) {
      for urlContext in URLContexts {
          let url = urlContext.url
          Auth.auth().canHandle(url)
      }
      // URL not auth related; it should be handled separately.
    }
}



@main
struct MyApp: App {
    
    @UIApplicationDelegateAdaptor(AppDelegate.self) var delegate

  
    // MARK: - UI Elements
    var body: some Scene {
        WindowGroup {
            NavigationView {
                ContentView()
            }
            
        }
    }
}
`

And here's what I am calling in the view when it crashes:

`    private func getVerificationCode(phoneNumber: String) {
        PhoneAuthProvider.provider()
          .verifyPhoneNumber(phoneNumber, uiDelegate: nil) { verificationID, error in
              // Sign in using the verificationID and the code sent to the user
          }
    }`

**Comment by atultw** [comment]
Ted, what worked for me was adding a new view such as OuterView wrapping the ContentView but within @main. 

Call firebase configure in MainView's onAppear and keep a flag like `firebaseLoaded`, then conditionally show ContentView if true. 

All Firebase APIs should be used only once the flag is true i.e. when ContentView loads. In other words the @main App and OuterView should not do anything with Firebase other than configure it.
